[josh1248] iP

josh1248 opened the pull request:
# Grok
> Grok' means to understand so thoroughly that the observer becomes a part of the process being observed - to merge, to blend, to intermarry, to lose personal identity in group experience. - Robert A. Heinlein

Understand and [be one with productivity](https://github.com/nus-cs2103-AY2425S1/ip/pull/225) - with the Grok chatbot. Makes productivity so easy that you embody productivity. 🥇
- 0% hassle
- Minimalist. ~~No extra junk.~~
- 100% smooth


**Features**
1. Ability to add and remove 3 types of todos - tasks, deadlines, and events
2. Seamless text-based storage that is persistent
3. Ability to mark and unmark todos as required


`Grok` just checks all the boxes:
- [x] Lightweight
- [x] Amazing
- [x] `Grok`

```Java
public class Grok {
    public static void launchApp() {
        You.setProductivity(110);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

josh1248 committed:
Grok.java: Add greeting
- Change chatbot name from Duke to Grok
- Add horizontal lines between messages

josh1248 committed:
Grok.java: Permit user input
- Echo user input, bot remains running if input is not 'bye'
- Exits bot if user input is 'bye'
- Indent bot input to separate from user input

josh1248 committed:
Grok.java: Store and list user commands
- Stores user commands if not 'list' or 'bye'
- Displays user commands if user input is 'list'

josh1248 committed:
Implement marking tasks as done or undone
- Task.java: Extract tasks as their own class
- Grok.java: Recognize new mark and unmark commands, with basic user
  input validation

josh1248 committed:
Permit different task subtypes for addition into tasklist
- Grok.java: Recognize "todo", "deadline", and "event" commands
  corresponding to different subtypes of tasks
- Task.java: Separate task subtypes to Todo, Deadline, and Event classes
  that inherit from the Task class

josh1248 committed:
runtest.sh: use shell scripts to semi-automate text UI testing

josh1248 committed:
Include user input validation when processing text commands

josh1248 committed:
Include user input validation when processing text commands

josh1248 committed:
Grok.java: Delete command for any subtype of task

josh1248 committed:
TextFileScanner.java: Create scaffold for CRUD on text files

josh1248 committed:
Grok.java: Increase robustness against non-valid user input

josh1248 committed:
TextFileReader.java: Create serialization method

josh1248 committed:
Convert naive string storage to DateTime

josh1248 committed:
Add Javadocs for more complicated methods

josh1248 committed:
Merge branch 'branch-Level-7'
Merge level 7 branch into master

josh1248 committed:
Fix text storage of DateTime

josh1248 committed:
Recognize date formatting in deadline and event tasks

josh1248 committed:
Organize related code into classes in OOP style

josh1248 committed:
Convert reader into OOP style code

josh1248 committed:
Declare task list as its own OOP class

josh1248 committed:
Initialize commands as an OOP class

josh1248 committed:
Revamp parser to OOP standards

josh1248 committed:
Fix parser error for mark and unmark

josh1248 committed:
Merge branch 'A-MoreOOP'

josh1248 committed:
Merge branch 'add-gradle-support'

josh1248 committed:
Add JUnit tests

josh1248 committed:
Remove duplicate .jar from shadowJar name

josh1248 committed:
Add JavaDoc documents to most classes and methods

josh1248 committed:
Fix coding standard violations

josh1248 committed:
Include Find command to filter by task description

josh1248 committed:
Add explantory notes to FindCommand

josh1248 committed:
Merge branch 'branch-A-JavaDoc'

josh1248 committed:
Merge branch 'branch-A-CodingStandard'

josh1248 committed:
Merge branch 'branch-Level-9'

zhou-colla commented on src/test/java/parser/ParserTest.java:
```diff
@@ -0,0 +1,46 @@
+
+import commands.ExitCommand;
+import exceptions.GrokInvalidUserInputException;
+import org.junit.jupiter.api.BeforeAll;
```
Would it looks neater if we separate the import statements for the import statements? 🤔

starrdye commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,39 @@
+     * Carry out the required actions of the command, including UI input, task list updating, and storage.
+     * Unifying under a single method allows for abstraction of the effects of the command from the caller.
+     */
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage);
```
Would it be better if we explains the functionalities of the parameters?

starrdye commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,152 @@
+package parser;
+
+import commands.*;
```
Is the wildcard necessary here?

josh1248 replied:
thank you for pointing this out - will resolve it!

starrdye commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,92 @@
+import taskList.TaskList;
+import tasks.Deadline;
+import tasks.Event;
+import tasks.Task;
```
I like how you sorted the import statements in order.

josh1248 committed:
Integrate checkstyle into Gradle

josh1248 committed:
Resolve existing coding style violations

josh1248 committed:
Merge branch 'branch-A-Checkstyle'

yastsc commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+     * @param isDone whether the deadline has been met or not
+     * @throws GrokInvalidUserInputException if user input is invalid - empty description, or non-ISO dates
+     */
+    public Deadline(String description, String due, Boolean isDone) throws GrokInvalidUserInputException {
```
Might be better to use the primitive type boolean instead of the reference type Boolean as argument type.

yastsc commented on src/main/java/ui/Ui.java:
All the method names in this class are very well named, and accurately describe what the method does, so good job!

yastsc commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,66 @@
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
+    }
+
```
Perhaps you can write more JavaDoc comments for the public methods?

yastsc commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,101 @@
+     * Converts a compliant text string from a text file into a list of tasks.
+     * @return an ArrayList of tasks.
+     */
+    public ArrayList<Task> parseTextStorage() {
```
loadTextStorage() may be a better name as the method doesn't just parse but also loads the stored tasks into an ArrayList that is returned.

yastsc reviewed (commented):
Overall, your naming of classes, methods and variables are really good! There seem to be only one or two method names that can be improved on. The coding standard is also mostly followed, so well done :)

josh1248 committed:
Add javafx tutorial structure

josh1248 committed:
Merge branch 'branch-Level-10'

josh1248 committed:
Transfer some GUI formatting to ui package

josh1248 committed:
Reroute controls

josh1248 committed:
Merge branch 'branch-Level-10'

josh1248 committed:
Fix checkstyle errors

josh1248 committed:
Add assertions to document assumptions in code

josh1248 committed:
Improve coverage of assertions in code

Previously, assertions mainly spanned in the task list as a means of out
of bounds checking.

This commit improves upon the power of assertions by enforcing certain
requirements before storage in text form is read and processed by the
program. Doing so prevents the damage that corrupted text files could
cause to the normal function of the program.

josh1248 committed:
Improve code quality of iP
Previously, some methods, such as storage parsing, spanned too many
lines (nearly 50).

This commit adds some coding standard remedies for this issue, breaking
long functions up to smaller ones that comprise 10 lines or less.

josh1248 committed:
Use streams for file processing
Previously, an iterative loop was used to handle and process different
lines in the list.

To reduce boilerplate behind this handling, the use of Java streams to
simplify list processing would be helpful.

josh1248 committed:
Split long methods
The coding standard cautions against excessive methods, and instead
recommends for methods to do one thing. Additionally, shorter methods
are less likely to violate SLAP.

Hence, this commit modifies existing methods in Storage to split the
text file parser into different methods to uphold this code quality.

josh1248 committed:
Merge pull request #3 from josh1248/branch-A-Assertions

Branch a assertions

josh1248 committed:
Merge branch 'master' into branch-A-Streams

josh1248 committed:
Merge branch 'master' into branch-A-CodeQuality

josh1248 committed:
Merge remote-tracking branch 'origin' into branch-A-CodingStandard

josh1248 committed:
Merge pull request #2 from josh1248/branch-A-CodingStandard

Split long methods

josh1248 committed:
Merge branch 'master' into branch-A-Streams

josh1248 committed:
Merge pull request #1 from josh1248/branch-A-Streams

Use streams for file processing

josh1248 committed:
Merge branch 'master' into branch-A-CodeQuality

josh1248 committed:
Merge pull request #4 from josh1248/branch-A-CodeQuality

Branch a code quality

josh1248 committed:
Prevent duplicate tasks from being added
Previously, duplicate tasks could be easily added into the task list,
which is undesirable as the user may be confused about why multiple
tasks with the same description are present.

A duplicate-checking method has been added to scan and ensure that no
task descriptions match before a task can be added. If a duplicate task
is added, addition of the task is aborted and the user is informed.

Streams are used as part of this feature to convey the stateless nature
of the duplicate checker method, allowing for multi-threaded use if
necessary.

josh1248 committed:
Merge pull request #5 from josh1248/C-DetectDuplicates

Prevent duplicate tasks from being added

josh1248 committed:
Improve robustness of task list tests

josh1248 committed:
Improve robustness of parser tests

josh1248 committed:
Fix import order in parser testing

josh1248 committed:
Merge branch 'branch-A-MoreTesting'

josh1248 committed:
Fix read/write issues when app is in a JAR file

josh1248 committed:
Fix user input for deadlines

josh1248 committed:
Include snapshot of product in use

josh1248 committed:
Update user guide

josh1248 committed:
Enhance user feedback

josh1248 committed:
Merge branch 'branch-A-UserGuide'

josh1248 committed:
Update user guide

josh1248 committed:
Update javadoc comments

josh1248 committed:
Update README.md
